pkg/storagev2/inmemstore/inmembatch.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package inmemstore
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+)
+
+func (s *Store) Batch(ctx context.Context) (storage.Batch, error) {
+	return &Batch{
+		mu:    new(sync.Mutex),
+		ctx:   ctx,
+		store: s,
+	}, nil
+}
+
+type Batch struct {
+	mu     *sync.Mutex
+	ctx    context.Context
+	put    []storage.Item
+	delete []storage.Key
+	store  *Store
+	done   bool
+}
+
+func (i *Batch) Put(item storage.Item) error {
+	if i.ctx.Err() != nil {
+		return i.ctx.Err()
+	}
+
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	for _, put := range i.put {
```

Comment:
So I see potential issues here. If the batch is big, the Put operation will be very slow as it has to iterate through all the ops every time. We should avoid this bottleneck in the Put operation.

What I would do is just have a map that is keyed on the key. That way duplicates will just be overwritten in the batch. This will shift the responsibility of correctly using the batch to the user.

Also, this can be done with just a single map.
```go
type batchOp struct {
    delete bool
    item storage.Item
}

type batch struct {
    ops map[string]batchOp
}
```
Even so only one op will be allowed for each key. If it is a duplicate op it is overwritten. This is the same as the behavior a user would see if he using the Store.

Reply:
fixed